Auto-/rollover with event sourcing

da-kami opened the pull request:
Finishes the rollover / auto-rollover implementation with event sourcing. This should finish our perpetual swap setup.

Fixes #1013 
Fixes https://github.com/itchysats/itchysats/issues/958
Fixes #466

thomaseizinger committed:
Introduce type alias for Completed<CollaborativeSettlement>

This makes it easier to introduce additional type parameters to
`Completed`.

thomaseizinger committed:
Introduce configurable error type for `Completed`

thomaseizinger committed:
Be more specific about handling the maker response timeout

Instead of failing the entire protocol after X amount of time, only
check whether our proposal got accepted by the maker within our timeout.
If it did, continue on. If it didn't fail the protocol. Not only should
this give us a much quicker response time in case the maker does not
response, it also more resilient in case rollover takes longer for
whatever reason.

Note that we still have timeout on the individual messages within
rollover anyway.

thomaseizinger committed:
Improve error handling when forwarding messages

thomaseizinger committed:
Make private what can be private

thomaseizinger committed:
Don't use spawn_fallible for the rollover task

The only error that can occur here is "Actor is disconnected". That can
only happen if we discard the rollover actor before the protocol
finishes. In that case, we are simply no longer interested in the result
and we can thus safely ignore this error.

To make sure that this thinking is not invalidated, we add an explicit
type annotation that enforces that we are not dropping an error here in
case someone ever changes the return value from the handler to NOT be a
`()` Ok value.

thomaseizinger committed:
Make internal messages private

thomaseizinger committed:
Remove duplicate error message

The error returned from `oracle::Actor` says exactly the same thing.

thomaseizinger committed:
Align error handling of `rollover_maker` sender with `rollover_taker`

thomaseizinger committed:
Use rollover noun instead of roll over

The only cases where we don't use the string "rollover" are when we
use the verb "to roll over".

da-kami committed:
Modernise (and fix) rollover after event sourcing refactor

- Reintroduce auto-rollover unit tests. The dummy CFD is now
constructed with the appropriate events.

- Separate rollover and auto-rollover decision logic. Being able to
rollover is independent from automatically checking whether we should
attempt to trigger rollover in the first place. The later depends on
time, so it's good to be able to make the rollover logic completely
independent of time.

- Re-introduce message to trigger rollover within auto-rollover. This
allows us to trigger rollover without the auto-rollover time based
check in the tests.

- Add rollover accept and reject actor tests. We should extend these
to verify that the CFD's internal details have actually changed (or
remained the same in the case of reject).

- Rename `RollOverProposed` to `RegisterRollover`. This message is
only used to tell the `maker_inc_connections::Actor` where to forward
messages from the taker about a particular rollover protocol for a
specific CFD. I think this name better reflects that.

Also the message is now sent from within the `rollove_maker::Actor`.
This introduces the need to pass yet another `MessageChannel` to its
constructor, but it allows us to further encapsulate all the things
related to rollover inside of the dedicated actor, further simplifying
the `maker_cfd::Actor`.

- Do not use catch-all pattern in process_manager

I think we should be explicit in what we want to handle and what we
want to ignore. Otherwise if we add a new event, the compiler will
assume that we don't need to handle it!

Co-authored-by: Lucas Soriano <lucas@comit.network>
Co-authored-by: Mariusz Klochowicz <mariusz@klochowicz.com>

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -522,26 +516,31 @@ impl Cfd {
+        self.can_roll_over()
+    }
+
+    fn can_roll_over(&self) -> Result<(), CannotRollover> {
```
🔧 To be consistent we could call this `is_rollover_possible`. Btw, I think we have settled on `rollover` vs `roll_over`, so we should fix that anyway.

da-kami replied:
True. I was unsure, because for e.g. collab settlement we use `can`. Happy to change it to `is` for concistency here. That might need another round of cleanup at some point.

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -591,8 +590,8 @@ impl Cfd {
-        if !self.can_roll_over() {
-            bail!("Start rollover only allowed when open")
+        if let Err(e) = self.can_roll_over() {
+            bail!(e)
```
🔧 I think you can just `?` here and below.

luckysori replied:
❓ Also, isn't it weird that internally we are using the non-actor-specific version of this API.

da-kami replied:
> ❓ Also, isn't it weird that internally we are using the non-actor-specific version of this API.

Both maker and taker share a set up rules, but the taker additionally checks the time since last rollover (I think the maker should not care about that). I wanted to keep the API consistent, but don't duplicate the code for the shared logic...

da-kami replied:
Your comment made me think again. The logic is actually incorrect at the moment I think. The rollover actor should check if it is possible to roll over, similar to the collaborative settlement actor. The cfd actor should not call `is_rollover_possible` for either party. For the taker the auto-rollover actor will call the taker specific one for decision making! I'll change that.

luckysori replied:
I think I agree!

luckysori commented on daemon/src/rollover_taker.rs:
```diff
@@ -190,10 +193,19 @@ impl xtra::Actor for Actor {
+                            order_id: self.cfd.id(),
+                            reason: e.into(),
+                        }
+                    }
```
I can look into this in a follow-up PR.

luckysori reviewed (approved):
LGTM

klochowicz commented on daemon/src/auto_rollover.rs:
```diff
@@ -73,68 +70,71 @@ where
-            .run();
+            let cfd = try_continue!(load_cfd(id, &mut conn).await);
+
+            if let Err(e) = cfd.can_auto_rollover_taker(OffsetDateTime::now_utc()) {
```
Can you check this with an event instead? I'd assume expired cfds have certain events appended to them? Otherwise, we really need to get back to mocking time to make this test work....

(additionally, both sides could check whether it got expired if we just analyse events? then at least we get rid of asymmetry, but it's a minor issue)

luckysori replied:
Because we have introduced the indirection of the `auto_rollover::Actor` sending a message to itself to spawn a `rollover_taker::Actor` we can actually directly send said message in the actor tests, circumventing this check.

klochowicz commented on daemon/src/rollover_taker.rs:
```diff
@@ -29,73 +27,88 @@ use futures::future;
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd(self.id, &mut conn).await?;
+
+        let (event, (rollover_params, dlc)) = cfd.start_rollover_taker()?;
```
would it be possible to only retrieve the `(rollover_params, dlc)` after acceptance? do we need it earlier?

luckysori replied:
I'll take a look tomorrow!

luckysori replied:
Good catch, I think it makes a lot more sense.

luckysori commented on daemon/src/auto_rollover.rs:
```diff
@@ -70,39 +69,40 @@ where
-            this.send_async_safe(Rollover { id }).await?;
+                    anyhow::Ok(())
+                }
+                .await
```
🔧 Let's add the "Cannot roll over" context to the result of this async block, so that any of the errors which could prevent the roll over have the appropriate context.

luckysori commented on daemon/src/auto_rollover.rs:
```diff
@@ -7,7 +7,6 @@ use crate::model::cfd::OrderId;
 use crate::oracle;
```
@da-kami Most of the changes to this file are because I used a version that @klochowicz and I worked on which is almost exactly the same as what you ended up doing, so not much to see here.

luckysori commented on daemon/src/lib.rs:
```diff
@@ -397,7 +398,6 @@ where
         )
         .create(None)
         .run();
-        std::mem::forget(auto_rollover_address); // leak this address to avoid shutdown
```
We are no longer leaking this because we can just keep it around and use it in the tests. Any reason why leaking was chosen over storing it in the actor system, @da-kami? (You may have been part of that work).

da-kami replied:
We did not store it in the actor system before because it was simply not needed (unused). But with using it in the tests it is. that is all good. I think it's fine to store it there for potential future use in production as well.

luckysori commented on daemon/src/maker_cfd.rs:
```diff
@@ -217,6 +217,13 @@ where
+                order_id: proposal.order_id,
+                address: rollover_actor_addr.clone(),
+            })
+            .await?;
```
@da-kami: I had to reintroduce this because we still need to tell the `maker_inc_connections::Actor` where to forward messages from the taker about the current rollover of this particular CFD. This sucks, but it may be solved down the line with QUIC's multiplexing which Thomas has been talking about.

da-kami replied:
hm... would it be possible to do this from the rollover actor instead? 
If not easily possible we can leave it here. Would it be worth adding a ticket describing that problem then?

luckysori replied:
I think moving it to the rollover actor makes more sense, let me try.

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -640,7 +633,7 @@ impl Cfd {
 
         Ok((
-            Event::new(self.id, CfdEvent::CollaborativeSettlementProposalAccepted),
+            Event::new(self.id, CfdEvent::RolloverAccepted),
```
Just a little copy-paste bug here.

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -655,6 +648,14 @@ impl Cfd {
         ))
     }
 
+    pub fn handle_rollover_accepted_taker(&self) -> Result<Event> {
```
I chose to add this for two reasons:

1. I want to be able to emit the accepted event on the taker side, so that we can update de UI.
2. I want to ensure that we only proceed with the rollover if the CFD's state still agrees with it.

da-kami replied:
Notes: 
ad. (1): This is actually more than we did before - the taker was not informed about accept before
ad. (2): Makes sense, but the time between sending the proposal and it being accepted should (usually) be within seconds. 

I think it is nice adding this step, but I don't think it is strictly necessary. Yes, going on with rollover if the Cfd is already final might cause weird side effects. I don't think it is such a big deal though, because once we apply the rollover we check again, so it should fail then. This *is* of course dangerous, because we already exchnaged parameters with the other party then, and the other party might not have noticed finality. In the end the `Cfd` would always be final though (once lock is spent it is spent...).

That said: happy to keep it, not sure if the state "helps" the user's understanding though. But that can be fixed by not showing fine-granular states in the UI.
If we want a separate state in the UI we might have to extend the state mapping in the projection actor.

luckysori replied:
> ad. (2): Makes sense, but the time between sending the proposal and it being accepted should (usually) be within seconds.

That is true, but this snippet should allow the user to know what's going on after the other party _accepts_. It should transition to `ContractSetup` (we could eventually introduce a dedicated state). Otherwise the UI would still show `Sent` or whatever we're doing atm.

From your comments I think we're okay to keep this, but thanks for thinking it through.

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -774,19 +775,19 @@ impl Cfd {
         // still able to apply it we should, otherwise we might produce inconsistent state with the
         // other party.
-        if let Err(e) = self.can_rollover() {
+        if let (Err(e), RolloverCompleted::Succeeded { .. }) = (self.can_rollover(), &completed) {
```
This was interesting. I noticed that we were hitting this `bail` when receiving a `RollOverCompleted::Failed` caused by failing on this very `can_rollover` check inside of a different method.

If the completion is failed or rejected, then we can safely emit the corresponding events. It's only in the case that we have succeeded in completing rollover _and_ rollover is no longer possible (e.g. because the commit transaction was found on the blockchain) that we should exit without emitting an event IMO.

da-kami replied:
Oh my, yes! No checks needed for failed/rejected. Good catch!

luckysori reviewed (changes requested):
Working on applying some of the commits of #966 onto this, I've noticed that we don't do any post-processing for rollover in the `process_manager` actor yet. That will need to be fixed for roll over to actually work.

luckysori committed:
Do not emit an event if rollover didn't take place

luckysori committed:
Fix warning message when failing to complete rollover

luckysori committed:
Align `rollover_maker` with `rollover_taker`

da-kami reviewed (commented):
Approved!

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -205,7 +204,7 @@ where
         let cfd = load_cfd(proposal.order_id, &mut conn).await?;
 
-        cfd.is_rollover_possible(OffsetDateTime::now_utc())?;
+        cfd.is_rollover_possible_maker()?;
```
This very much looks like a pre-check of a command that the maker should be running against its CFD.

luckysori replied:
Good point, I think we need to remove this.

luckysori replied:
Okay, this was removed and done the way you suggest here.

klochowicz commented on daemon/src/projection.rs:
```diff
@@ -428,26 +414,22 @@ impl Cfd {
+                Role::Taker => (CfdState::OutgoingRollOverProposal, vec![]),
+            },
+            RolloverAccepted => {
+                // TODO: Might wanna introduce a separate state there...?
```
Indeed - especially if we decide to allow collab settlement (ie. closing the cfd) during the rollover phase after all, as @da-kami  suggested. 
It's an UI-only state, so that should be trivial to add (just aligning the frontend to handle it correctly).
can be done later though.

luckysori replied:
See https://github.com/itchysats/itchysats/issues/1037.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -1816,4 +1822,345 @@ mod tests {
+    }
+
+    impl Event {
+        fn dummy_open(event_id: BitMexPriceEventId) -> Vec<Self> {
+            vec![
+                Event {
+                    timestamp: Timestamp::now(),
+                    id: Default::default(),
+                    event: CfdEvent::ContractSetupStarted,
+                },
+                Event {
+                    timestamp: Timestamp::now(),
+                    id: Default::default(),
+                    event: CfdEvent::ContractSetupCompleted {
+                        dlc: Dlc::dummy(Some(event_id)),
+                    },
+                },
+                Event {
+                    timestamp: Timestamp::now(),
+                    id: Default::default(),
+                    event: CfdEvent::LockConfirmed,
+                },
+            ]
+        }
```
This is a great first step!

One advantage of an event-sourcing based system is that we can invest some effort into a test harness that constructs us the different events. I.e. if we change this to something like:

```rust
impl Event {
	fn dummy_contract_setup_started() {}
	fn dummy_contract_setup_completed() {}
	// etc
}
```

Then the code for `Cfd::dummy_open` could read like:

```rust
[
	Event::dummy_contract_setup_started(),
	Event::dummy_contract_setup_completed(),
]
.into_iter()
.fold(cfd, Cfd::apply)
```

Which reduces a layer of indirection and makes these events more reusable for other tests! Reusing the same events might help us test stuff within `monitor` for example!

da-kami replied:
Definitely! The original purpose of this PR was to get the unit tests for auto-rollover back in and I tried to do that in the fastest way. 
I think we should generally invest time into a test harness for unit tests for `Cfd`. What we do for auto-rollover could and should be done for other features as well! We can unit test the complete event logic in the `Cfd`. 

I'd propose to discuss this together once you are back and ramp up a simple harness that can then be steadily extended (simliar to how we handled the actor tests...).

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -211,14 +211,20 @@ pub enum SettlementKind {
+    TooRecent,
+    #[error("Cfd does not have a dlc")]
+    NoDlc,
+}
```
If we don't end up sending these errors around as values, I think I wouldn't bother modelling them!

da-kami replied:
We went back and forth with these "cases" (`TooRecent` is not even a real error case, because you just can't rollover but there is nothing erroneous about that...).

For me, returning specific error types greatly helps the understanding of the logic, because I immediately see what is the problem. Furthermore, the caller can decide to react differently. Maybe there is a better way of modelling this, that does not actually build on errors. I would not invest the time to change them know though.

luckysori replied:
I'm keeping them around for the time being.

thomaseizinger commented on daemon/src/rollover_taker.rs:
```diff
@@ -128,9 +117,9 @@ impl Actor {
             receiver,
             (self.oracle_pk, announcement),
-            rollover_params,
+            todo!("plug in rollover params from self once we have them"),
```
Not particularly clean having these in here within the commit. Not sure if you plan on squashing?

luckysori replied:
I think the idea is to squash most of this, yep.

da-kami replied:
@luckysori is planning to squash all the commits together. We kept them for now to make reviewing easier. We should have probably stated that clearer 😬

klochowicz commented on daemon/src/rollover_maker.rs:
```diff
@@ -256,6 +238,54 @@ impl Actor {
+    }
+
+    async fn stopping(&mut self, ctx: &mut Context<Self>) -> KeepRunning {
+        let address = ctx.address().expect("acquired own actor address");
```
nit: `"self to be alive" , just like in other case above?

luckysori replied:
I always grep the codebase and copy paste the first hit lol.

thomaseizinger commented on daemon/src/auto_rollover.rs:
```diff
@@ -21,12 +23,12 @@ use xtra::Actor as _;
     db: sqlx::SqlitePool,
     oracle_pk: schnorrsig::PublicKey,
     projection_actor: Address<projection::Actor>,
+    process_manager_actor: Address<process_manager::Actor>,
```
Nit: Would prefer the naming to omit `actor` for brevity.

thomaseizinger commented on daemon/src/auto_rollover.rs:
```diff
@@ -127,36 +128,49 @@ where
+            .send(process_manager::Event::new(event.clone()))
+            .await?
+        {
+            tracing::error!("Sending event to process manager failed: {:#}", e);
```
My current view is that the `ProcessManager` failing is a hard error and should always abort the current function. If the caller wants to only log that result, that is up to them!

luckysori replied:
Yeah, I agree with you.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -737,22 +738,24 @@ impl Cfd {
+        // Daniel says: I don't think the version check will help in addition to what we check in
+        // can_rollover. What difference would it make? If we are completed rollover and are
+        // still able to apply it we should, otherwise we might produce inconsistent state with the
+        // other party.
```
Looking forward to the day where we have entire conversations incode  comments :rofl:

thomaseizinger replied:
Thomas says: For now I would make it a hard requirement to not allow any concurrency within these commands. If we find this restrictive, we can always relax this later. If the version we started rollover with is not the same as here, all kinds of state changes could have happened and reasoning about this is **not** trivial. Best to fail early here.

luckysori replied:
I will do a version check here.

da-kami replied:
Daniel says: Not sure it's worth it, you will have to pull the version through into the rollover actor and whatnot... But feel free to do it :D

🤪

luckysori replied:
I will contradict myself. Let's introduce this version check (or not) in a separate PR. Here's an issue where we can continue to discuss this particular case: https://github.com/itchysats/itchysats/issues/1038.

klochowicz commented on daemon/tests/happy_path.rs:
```diff
@@ -191,6 +191,43 @@ async fn force_close_an_open_cfd() {
+
+    assert_next_state!(CfdState::ContractSetup, maker, taker, order_id);
+    assert_next_state!(CfdState::Open, maker, taker, order_id);
+}
```
in a follow-up, we should really finish this test by force-closingw this cfd using a different dlc & event id, to confirm that the internals have changed (for that we'll need to extend the test framework slightly, but that should be quite easy :) )

luckysori replied:
Agreed as discussed earlier. I'll open an issue once this merges.

luckysori replied:
See https://github.com/itchysats/itchysats/issues/1036.

klochowicz reviewed (approved):
great job!!! really excited to see rollover working 🎉

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -598,13 +599,48 @@ impl Cfd {
+        ))
+    }
+
+    pub fn receive_rollover_proposal(self, proposal: RolloverProposal) -> Result<Event> {
```
I am liking this!

luckysori replied:
Good to get some more confirmation, because I wasn't sure if these state checking methods that just emit events were what we wanted necessarily.

thomaseizinger commented on daemon/src/rollover_maker.rs:
```diff
@@ -47,102 +47,94 @@ pub struct RolloverFailed {
+                    .await;
+            }
+            Err(e) => {
+                tracing::warn!(%order_id, "Failed to report completion of collab settlement: {:#}", e)
```
Copy paste error in log message :)

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -191,6 +191,43 @@ async fn force_close_an_open_cfd() {
+
+    let (taker_cfd, maker_cfd) = next_cfd(taker.cfd_feed(), maker.cfd_feed()).await.unwrap();
+    assert_eq!(taker_cfd.state, CfdState::OutgoingRollOverProposal);
+    assert_eq!(maker_cfd.state, CfdState::IncomingRollOverProposal);
```
With a bit more macro trickery to accept different syntaxes, we should be able to create a version of `assert_next_state` that allows us to assert on different events for maker / taker.

See https://doc.rust-lang.org/rust-by-example/macros/overload.html for example.

luckysori replied:
I briefly added something like what you suggest, but less elegant. I'll take a look at that link!

thomaseizinger commented on daemon/src/process_manager.rs:
```diff
@@ -163,14 +163,29 @@ impl Actor {
                 tracing::info!(%txid, "Commit transaction published");
             }
+            RolloverCompleted { dlc } => {
+                tracing::info!("Rollover complete, monitoring new txs and oracle events");
```
I think we should split this into:

1. Logging here that rollover completed
2. Logging in `monitor` that we are starting to monitor a new DLC
3. Logging in `oracle` that we are starting to monitor a new attestation

thomaseizinger commented on daemon/src/rollover_maker.rs:
```diff
@@ -98,7 +98,7 @@ impl Actor {
         let mut conn = self.db.acquire().await?;
         let cfd = load_cfd(self.proposal.order_id, &mut conn).await?;
 
-        let event = cfd.receive_rollover_proposal(self.proposal.clone())?;
```
I think this highlights how useless the data in `RolloverProposal` currently is :joy:

luckysori replied:
Yep, we want to get rid of the timestamp. The order id is necessary at some level, but maybe the `RolloverProposal` itself can omit it?

da-kami replied:
Depends on what level. When sending from taker to maker we must include the id. 
But within the context for a rollover actor or the `Cfd` we know the id anyway...

thomaseizinger replied:
- On the wire we certainly need the ID.
- Once we loaded the CFD from the DB, a parameter-less command should be enough to move forward.

thomaseizinger reviewed (commented):
Stellar work, some non-blocking comments!

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -982,11 +1089,22 @@ impl Cfd {
+            RolloverStarted => {
+                self.during_rollover = true;
+            }
+            RolloverAccepted => {} /* TODO: It's weird we have no state change for applying an */
```
@thomaseizinger @da-kami: I'm removing this TODO because I hate all code TODOs, but the point stands. Is this a bad sign? The `RolloverAccepted` was only introduced because a different aggregate would care (in `projection`). Maybe that's fine.

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -982,11 +1089,22 @@ impl Cfd {
+                self.during_rollover = false;
+            }
+            RolloverFailed { .. } => {
+                // TODO: Deal with failed rollover
```
@da-kami: Also removing this TODO. What are we not doing that we should be doing? Then I can open an issue.

luckysori committed:
Expose strongly typed rollover errors

luckysori committed:
Re-order assert_next_state macro arguments

luckysori committed:
Add optional argument to assert_next_state macro

So that it can also be used when the maker and taker are expected to
be in different states.
